lib/pbench/server/database/models/audit.py

```diff
@@ -0,0 +1,376 @@
+    total success."""
+    WARNING = enum.auto()
+
+
+class AuditReason(enum.Enum):
+    """A high-level 'reason' for a `FAILURE` or `WARNING` audit record. These
+    records should also contain a 'message' string in the 'attributes' to
+    provide more detail."""
+
+    """Permission denied."""
+    PERMISSION = enum.auto()
+
+    """Internal failure."""
+    INTERNAL = enum.auto()
+
+    """Consistency failure. For example, a corrupted tarball upload."""
+    CONSISTENCY = enum.auto()
+
+
+class Audit(Database.Base):
+    """
+    A framework to store Pbench audit records. These will track server
+    configuration changes as well as every mutation of user-visible data and
+    attribute each to a specific user and operation. Mutations attributed to
+    "the server" independent of a specific authenticated user will be attached
+    to a "BACKGROUND" user.
+
+    Architecturally, the SQL table may be used as a "front end" for a more
+    expensive audit repository (e.g., Elasticsearch). For example, we might
+    periodically bulk-replicate these rows as JSON data into an Elasticsearch
+    index and wipe the table.
+
+    Columns:
+        id: Generated unique ID of table row
+        root_id: Reference to the id of the operation's BEGIN audit record to
+            enable following an operation with multiple records.
+        name: Provide an optional name for the operation, like "upload" or
+            "index".
+        operation: The operation performed
+        object_type: The type of resource
+        object_id: The object on which an operation was performed
+        object_name: The name of the object (more meaningful than ID if the
+            resource has been deleted)
+        user_id: The identity of the user performing an operation
+        user_name: The name of the user (more meaningful than ID if the
+            user has been deleted)
+        status: The status of the operation
+        attributes: A JSON document of attributes ("message" key is a common
+            convention but others may be included for clarity)
+        timestamp: The date/time the operation was performed
+    """
+
+    __tablename__ = "audit"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    root_id = Column(Integer, nullable=True)
+    name = Column(String(128), nullable=True)
+    operation = Column(Enum(OperationCode), index=True, nullable=False)
+    object_type = Column(Enum(AuditType), index=False, nullable=True)
+    object_id = Column(String(128), index=False, nullable=True)
+    object_name = Column(String(256), nullable=True)
+    user_id = Column(String(128), index=True, nullable=True)
+    user_name = Column(String(256), index=False, nullable=True)
+    status = Column(Enum(AuditStatus), index=True, nullable=False)
+    reason = Column(Enum(AuditReason), nullable=True)
+    attributes = Column(JSON, index=False, nullable=True)
+    timestamp = Column(TZDateTime, nullable=False, default=TZDateTime.current_time)
```

Comment:
`index=True` is an optimization to tell PostgreSQL to build an index for that column. The `primary_key` is always indexed, but others can be as well.

It's quite possible that it's not worth building an index for any of these, but commonly searched columns will benefit from a dedicated index.

I'm pretty sure the default is to not index (the converse wouldn't make any sense), and this can be changed dynamically without affecting the schema (just performance). So, yeah; a lot of this was boilerplate copied and quickly edited from other schemas before I really thought about it.

I should probably drop indexing from most of the columns; maybe all of them. Hmm.